File: blocksandstuff-blocks/src/main/java/org/everbuild/blocksandstuff/blocks/BlockPlacementRuleRegistrations.java
package org.everbuild.blocksandstuff.blocks;

import net.minestom.server.MinecraftServer;
import net.minestom.server.instance.block.Block;
import net.minestom.server.instance.block.BlockManager;
import org.everbuild.blocksandstuff.blocks.group.block.BlockGroup;
import org.everbuild.blocksandstuff.blocks.group.placement.PlacementGroup;
import org.everbuild.blocksandstuff.blocks.group.VanillaPlacementRules;

public class BlockPlacementRuleRegistrations {
    public static void registerDefault() {
        register(VanillaPlacementRules.ALL.toArray(new PlacementGroup[0]));
    }

    public static void register(PlacementGroup... blockGroups) {
        BlockManager blockManager = MinecraftServer.getBlockManager();

        for (PlacementGroup placementGroup : blockGroups) {
            BlockGroup blockGroup = placementGroup.getBlockGroup();
            for (Block block : blockGroup.allMatching()) {
                blockManager.registerBlockPlacementRule(placementGroup.createRule(block));
            }
        }
    }
}

File: blocksandstuff-blocks/src/main/kotlin/org/everbuild/blocksandstuff/blocks/BlockPlacementRuleRegistrations.kt
package org.everbuild.blocksandstuff.blocks

import net.minestom.server.MinecraftServer
import net.minestom.server.instance.block.Block
import org.everbuild.blocksandstuff.blocks.group.VanillaPlacementRules
import org.everbuild.blocksandstuff.blocks.group.placement.PlacementGroup

object BlockPlacementRuleRegistrations {
    private val logger = MinecraftServer.LOGGER

    @JvmStatic
    fun registerDefault() {
        registerDefault(listOf())
    }

    @JvmStatic
    fun registerDefault(exclude: List<Block>) {
        register(exclude, *VanillaPlacementRules.ALL.toTypedArray<PlacementGroup>())
    }

    @JvmStatic
    fun register(vararg blockGroups: PlacementGroup) {
        register(listOf(), *blockGroups)
    }

    @JvmStatic
    fun register(exclude: List<Block>, vararg blockGroups: PlacementGroup) {
        val blockManager = MinecraftServer.getBlockManager()
        var count = 0

        for (placementGroup in blockGroups) {
            val blockGroup = placementGroup.blockGroup
            for (block in blockGroup.allMatching()) {
                if (exclude.contains(block)) continue
                count++
                blockManager.registerBlockPlacementRule(placementGroup.createRule(block))
            }
        }

        logger.info("Registered $count block placement rules")
    }
}
